build-system/tasks/visual-diff.js

```diff
@@ -384,56 +390,142 @@ async function snapshotWebpages(percy, page, webpages, config) {
  */
 async function verifyCssElements(page, url, forbiddenCss, loadingIncompleteCss,
   loadingCompleteCss) {
-  // Wait for loader dot to be hidden.
-  await page.waitForSelector('.i-amphtml-loader-dot', {
-    hidden: true,
-    timeout: CSS_SELECTOR_TIMEOUT_MS,
-  }).catch(() => {
-    log('fatal', colors.cyan(url),
-        `still has the AMP loader dot after ${CSS_SELECTOR_TIMEOUT_MS} ms`);
-  });
+  // Begin by waiting for all loader dots to disappear.
+  await waitForLoaderDot(page, url);
 
   if (forbiddenCss) {
     for (const css of forbiddenCss) {
       if (await page.$(css) !== null) {
-        log('fatal', colors.cyan(url), 'has forbidden CSS element',
-            colors.cyan(css));
+        log('fatal', colors.cyan(url), '| The forbidden CSS element',
+            colors.cyan(css), 'exists in the page');
       }
     }
   }
 
   if (loadingIncompleteCss) {
+    log('verbose', 'Waiting for invisibility of all:',
+        colors.cyan(loadingIncompleteCss.join(', ')));
     for (const css of loadingIncompleteCss) {
-      log('verbose', `Waiting for invisibility of ${css}`);
-      await page.waitForSelector(css, {
-        hidden: true,
-        timeout: CSS_SELECTOR_TIMEOUT_MS,
-      }).catch(() => {
-        log('fatal', colors.cyan(url), 'still has CSS element',
-            colors.cyan(css), `after ${CSS_SELECTOR_TIMEOUT_MS} ms`);
-      });
+      if (!(await waitForElementVisibility(page, css, {hidden: true}))) {
+        log('fatal', colors.cyan(url), '| An element with the CSS selector',
+            colors.cyan(css),
+            `is still visible after ${CSS_SELECTOR_RETRY_MS} ms`);
+      }
     }
   }
 
   if (loadingCompleteCss) {
+    log('verbose', 'Waiting for existence of all:',
+        colors.cyan(loadingCompleteCss.join(', ')));
+    for (const css of loadingCompleteCss) {
+      if (await !waitForSelectorExistence(page, css)) {
+        log('fatal', colors.cyan(url), '| The CSS selector', colors.cyan(css),
+            'does not match any elements in the page');
+      }
+    }
+
+    log('verbose', 'Waiting for visibility of all:',
+        colors.cyan(loadingCompleteCss.join(', ')));
     for (const css of loadingCompleteCss) {
-      log('verbose', 'Waiting for visibility of', css);
-      await page.waitForSelector(css, {
-        visible: true,
-        timeout: CSS_SELECTOR_TIMEOUT_MS,
-      }).catch(() => {
-        log('fatal', colors.cyan(url), 'does not yet have CSS element',
-            colors.cyan(css), `after ${CSS_SELECTOR_TIMEOUT_MS} ms`);
-      });
+      if (!(await waitForElementVisibility(page, css, {visible: true}))) {
+        log('fatal', colors.cyan(url), '| An element with the CSS selector',
+            colors.cyan(css),
+            `is still invisible after ${CSS_SELECTOR_RETRY_MS} ms`);
+      }
     }
   }
+
+  // Again, wait for all loader dots to disappear. Sometimes components are
+  // still loading images, videos, etc.
+  await waitForLoaderDot(page);
```

Comment:
Curious: Are there cases where the previous call succeeds, and this call actually does some more waiting?

Reply:
I measured the timing of both having these lines in and not having them in (code: https://gist.github.com/danielrozenberg/44edd76f8f6c025f6cfcf967762d1581 and https://gist.github.com/danielrozenberg/d2dbf24403d1bc8d4ece61a5496b91b6)

On an 100 repetitions A/B test, the average call to verifyCssElements with/without the code is 9.09ms/8.59ms which I would say is practically insignificant.

However, I noticed that the loader dot was never found in the 200 repetitions on the 2nd part, so I'll remove this code. It might be useful again when we reactivate ABE, so I'll keep that in mind.